src/dynamo0p3.py

```diff
@@ -1301,13 +1301,15 @@ def __init__(self, alg_invocation, idx):
                         # 1 to the vector size which is what we
                         # require in our Fortran code
                         for idx in range(1, halo_field.vector_size+1):
-                            exchange = DynHaloExchange(
-                                halo_field, parent=loop, vector_index=idx,
-                                inc=inc)
+                            exchange = DynHaloExchange( halo_field,
+                                                        parent=loop.parent,
+                                                        vector_index=idx,
+                                                        inc=inc)
                             loop.parent.children.insert(loop.position,
                                                         exchange)
                     else:
-                        exchange = DynHaloExchange(halo_field, parent=loop,
+                        exchange = DynHaloExchange(halo_field,
```

Comment:
Yes, I found this bug when adding tests in this issue. I could argue that it is therefore now covered, as one of the new tests will fail if the bug is reintroduced. However, I think you are right that there should be a specific test that halo exchanges have the correct parent set up. I will therefore add a test.